__author__ = "Arielle R. Munters"
__copyright__ = "Copyright 2022, Arielle R. Munters"
__email__ = "arielle.munters@scilifelab.uu.se"
__license__ = "GPL-3"

import pandas as pd
from snakemake.utils import validate
from snakemake.utils import min_version

from hydra_genetics.utils.resources import load_resources
from hydra_genetics.utils.samples import *
from hydra_genetics.utils.units import *
from hydra_genetics import min_version as hydra_min_version

hydra_min_version("0.15.0")
min_version("6.8.0")

### Set and validate config file


configfile: "config.yaml"


validate(config, schema="../schemas/config.schema.yaml")
config = load_resources(config, config["resources"])
validate(config, schema="../schemas/resources.schema.yaml")


### Read and validate samples file

samples = pd.read_table(config["samples"], dtype=str).set_index("sample", drop=False)
validate(samples, schema="../schemas/samples.schema.yaml")

### Read and validate units file

units = (
    pandas.read_table(config["units"], dtype=str)
    .set_index(["sample", "type", "flowcell", "lane", "barcode"], drop=False)
    .sort_index()
)
validate(units, schema="../schemas/units.schema.yaml")


def get_gvcfs(units: pandas.DataFrame) -> typing.List[str]:
    """
    function used to extract all gvcf files found in units.tsv
    Args:
        units: DataFrame generate by importing a file following schema definition
               found in pre-alignment/workflow/schemas/units.schema.tsv
    Returns:
        List of strings with all gvcf file names and path
    """
    return get_units_column(units, "gvcf")


def get_units_column(units: pandas.DataFrame, column: str) -> typing.List[str]:
    """
    extract a column from units.tsv
    Args:
        units: DataFrame generated by importing a file following schema definition
               found in workflow/schema/units.schema.yaml
        column: the name of the requested column
    Returns:
        List of strings representing the content of the requested column, or an
        empty list if the column does not exist, or if all values are missing/empty
    """
    if column not in units.columns:
        return []
    return list(set(units[column][units[column].notna()]))


def generate_read_group_star(wildcards):
    return "--outSAMattrRGline  ID:{} SM:{} PL:Illumina PU:{} LB:{}".format(
        "{}_{}".format(wildcards.sample, wildcards.type),
        "{}_{}".format(wildcards.sample, wildcards.type),
        "{}_{}".format(wildcards.sample, wildcards.type),
        "{}_{}".format(wildcards.sample, wildcards.type),
        "{}_{}".format(wildcards.sample, wildcards.type),
    )


### Set wildcard constraints
wildcard_constraints:
    sample="|".join(samples.index),
    type="N|T|R",


def compile_output_list(wildcards):
    output_files = [] #["Results/MultiQC_R.html"]s
    output_files.append(
        [
            "Results/%s_%s/%s_%s.%s" % (sample, type, sample, type, suffix)
            for sample in get_samples(samples)
            for type in get_unit_types(units, sample)
            for suffix in [
                # "bam",
                # "bam.bai",
#                "arriba.fusions.tsv",
#                "arriba.pdf",
                "merged.vcf.gz",
                "merged.vcf.gz.tbi",
                "config.yaml",
            ]
        ]
    )
    # output_files.append(
    #     [s
    #         "Results/%s_%s_summary.xlsx" % (sample, type)
    #         for sample in get_samples(samples)
    #         for type in get_unit_types(units, sample)
    #     ]
    # )
    return output_files
